### DISCUSSION TITLE
wanted: stronger typing for pytest fixtures

### DISCUSSION THREAD

**Discussion by bukzor** [microsoft/pyright#7135] [Q&A]
The pytest system uses a dependency-injection system which pyright can't statically analyze. In brief, the system matches objects defined in a sibling `conftest.py` to test arguments of the same name. This setup represents a very common test suite: (no problems here, yet)

`conftest.py`
```python
import pytest


class EffFactory:
    def __call__(self):
        return "f"


@pytest.fixture
def eff() -> EffFactory:
    return EffFactory()
```

`my_test.py`
```python
from .conftest import EffFactory


class TestMy:
    def test1(self, eff: EffFactory) -> None:
        assert eff() == "f"
```

The system works, and is well-typed:

```console
$ python --version; pyright --version; pytest --version
Python 3.12.1
pyright 1.1.348
pytest 7.4.4

$ pyright .
0 errors, 0 warnings, 0 informations

$ pytest .
$ pytest .
my_test.py .                                                             [100%]
============================== 1 passed in 0.01s ===============================
```

Settings used:

`pyproject.toml`
```toml
$ cat pyproject.toml
[tool.pytest.ini_options]

[tool.pyright]
pythonPlatform = "Linux"
pythonVersion = "3.12"

typeCheckingMode = "strict"

# turn on all the checks not already enabled by "strict":
reportCallInDefaultInitializer = "error"
reportImplicitStringConcatenation = "error"
reportMissingSuperCall = "error"
reportPropertyTypeMismatch = "error"
reportUninitializedInstanceVariable = "error"
reportUnnecessaryTypeIgnoreComment = "error"
reportUnusedCallResult = "none"  # too noisy
```

The situation that I would like to improve (because I keep creating bugs) is that when a refactor changes the type of the fixture but not the test, pyright (of course) doesn't notice:

If a refactor changes the type in the fixture but not the test, pyright (of course) won't be able to notice the problem. For example:

`conftest.py`
```python
import pytest


class EffFactory:
    def __call__(self):
        return "f"


@pytest.fixture
def eff() -> str:
    return "f"
```

Then pyright is happy but there's a major type error:

```console
$ pyright .
0 errors, 0 warnings, 0 informations

$ pytest . | cat
my_test.py F                                                             [100%]

=================================== FAILURES ===================================
_________________________________ TestMy.test1 _________________________________

self = <python.my_test.TestMy object>, eff = 'f'

    def test1(self, eff: EffFactory) -> None:
>       assert eff() == "f"
E       TypeError: 'str' object is not callable

my_test.py:6: TypeError
============================== 1 failed in 0.02s ===============================
```

#### Problem Statement
Would you help me find some practical improvement for this situation? (Dropping pytest isn't feasible.)

#### Attempted Solutions

This seems to "work", but I suspect this has entered the realm of unintended behavior, and full coverage would require each and every fixture usage to contain this extra assertion:

`my_test.py`
```python
from typing import assert_type

from . import conftest
from .conftest import EffFactory


class TestMy:
    def test1(self, eff: EffFactory) -> None:
        assert_type(eff, "type(conftest.eff())")
        assert eff() == "f"
```

```console
$ pyright . | sed "s,$PWD,.,g"
./my_test.py
  ./my_test.py:9:21 - error: "assert_type" mismatch: expected "str" but received "EffFactory" (reportAssertTypeFailure)
1 error, 0 warnings, 0 informations
```

If I try the same hack on the argument annotation, pyright (of course) complains:

```
  ./my_test.py:8:27 - error: type() call should not be used in type annotation
    Use Type[T] instead (reportInvalidTypeForm)
  ./my_test.py:8:27 - error: Call expression not allowed in type expression (reportInvalidTypeForm)
```

If, again hypothetically, I had some inverse to `Type` (perhaps `Call` or `Return`), then I could handle this pretty well:

```python
from typing import Call

from . import conftest


class TestMy:
    def test1(self, eff: Return[conftest.eff]) -> None:
        assert eff.f() == "f"
```

Hopefully this gives you a clear idea of my goals. I hope your ideas are more practical than mine :) Thanks in advance for your time, reading this far.

**Comment by erictraut** [comment]
Unfortunately, I don't have any great ideas for improving type checking for pytest fixtures. They use a technique that involves runtime introspection and parameter injection — techniques that are too dynamic for the Python static type system to handle.

A (small) silver lining is that bugs in tests are typically pretty easy to catch because all code paths in a test are typically executed at test time, and all results of a test tend to be validated.

I personally try to avoid pytest its fixture mechanism because it not only doesn't work with static type checking, but I also find such tests to be really hard to read and understand. There's too much under-the-covers "magic" going on for me. But it sounds like moving away from pytest isn't an option for you.

**Reply by debonte** [reply]
Pylance adds some pytest support that is not available in pyright. For example, when you hover over an unannotated test parameter, Pylance finds the corresponding fixture and displays appropriate hover text for it:

![image](https://github.com/microsoft/pyright/assets/12820898/fba4d65a-4f47-4bbb-9591-78e5ac077fbb)

So if you want better pytest support, I'd suggest using Pylance and filing issues about enhancements that you'd find useful. For example, there's already an open enhancement item about providing completions for fixture objects within test code -- https://github.com/microsoft/pylance-release/discussions/4388.

cc: @rchiodo
